internal/service/alarms/internal/db/repo/alarms_repository.go

```diff
@@ -137,3 +149,229 @@ func (ar *AlarmsRepository) UpsertAlarmDefinitions(ctx context.Context, records
+				"AlarmName", "AlarmDefinitionID", "ProbableCauseID",
+				"ObjectTypeID",
+			}))...),
+		sm.From(m.TableName()),
+		sm.Where(
+			psql.Group(
+				psql.Quote(dbTags["AlarmName"]),
+				psql.Quote(dbTags["ObjectTypeID"])).
+				In(getGetAlertNameAndObjectTypeID(am, clusterIDToObjectTypeID)...), // Dynamically pass the pairs
+		),
+	)
+
+	sql, params, err := query.Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build alarm definitions query when processing AM notification: %w", err)
+	}
+
+	return utils.ExecuteCollectRows[models.AlarmDefinition](ctx, ar.Db, sql, params)
+}
+
+func getGetAlertNameAndObjectTypeID(am *api.AlertmanagerNotification, clusterIDToObjectTypeID map[uuid.UUID]uuid.UUID) []bob.Expression {
+	b := make([]bob.Expression, 0, len(am.Alerts))
+	for _, alert := range am.Alerts {
+		name := alertmanager.GetAlertName(*alert.Labels)
+		ObjectTypeID := clusterIDToObjectTypeID[alertmanager.GetClusterID(*alert.Labels)]
+		b = append(b, psql.ArgGroup(name, ObjectTypeID))
+	}
+
+	return b
+}
+
+// ArchiveResolvedAlarmEventRecords Delete resolved and move to archive in one transaction
+func (ar *AlarmsRepository) ArchiveResolvedAlarmEventRecords(ctx context.Context) error {
+	return pgx.BeginFunc(ctx, ar.Db, func(tx pgx.Tx) error { //nolint:wrapcheck
+		m := models.AlarmEventRecord{}
+		dbTags := utils.GetAllDBTagsFromStruct(m)
+		query := psql.Delete(
+			dm.From(m.TableName()),
+			dm.Where(psql.Quote(dbTags["AlarmStatus"]).EQ(psql.Arg(api.Resolved))),
+			dm.Returning(dbTags.Columns()...),
+		)
+
+		sql, params, err := query.Build()
+		if err != nil {
+			return fmt.Errorf("failed to build delete alarm record query: %w", err)
+		}
+
+		// Run query to delete
+		rows, _ := tx.Query(ctx, sql, params...)
+		records, err := pgx.CollectRows(rows, pgx.RowToStructByNameLax[models.AlarmEventRecord])
+		if err != nil {
+			return fmt.Errorf("failed to delete alarm records: %w", err)
+		}
+
+		// Move deleted data to archived table which is identical to the original. Archived data will later be removed with cronjob.
+		if len(records) > 0 {
+			// Build queries for each record
+			sql, params, err := buildAlarmEventRecordUpsertQuery(records, m.ArchiveTableName())
+			if err != nil {
+				return fmt.Errorf("failed to build query for event upsert: %w", err)
+			}
+
+			_, err = tx.Exec(ctx, sql, params...)
+			if err != nil {
+				return fmt.Errorf("failed to execute upsert query: %w", err)
+			}
+			slog.Info("Successfully archived alarm event record records", "records", len(records))
+		}
+
+		return nil
+	})
+}
+
+// ResolveNotificationIfNotInCurrent find and only keep the alerts that are available in the current payload
+func (ar *AlarmsRepository) ResolveNotificationIfNotInCurrent(ctx context.Context, am *api.AlertmanagerNotification) error {
+	m := models.AlarmEventRecord{}
+	dbTags := utils.GetAllDBTagsFromStruct(m)
+	query := psql.Update(
+		um.Table(m.TableName()),
+		um.SetCol(dbTags["AlarmStatus"]).ToArg(api.Resolved),
```

Comment:
As per our discussion about the clearing type of Alarms, change the perceived_severity to CLEARED here. This corresponds to the IMS-Initiated Alarm Clear section for automatically cleared alarms.

Also, you should also change the `alarmChangedTime` field. The specification says that this should change if any of the Alarm Record fields changes. So, this is not something that should be sent by the AlertManager but something is completely under our control.

There are 2 ways to change an Alarm Record, when it is acknowledged and when it is cleared. So, similar to what we do in the PATCH Alarm Record, save the current timestamp here since your are clearing the alarm.

Reply:
/hold

true. updating this now

Reply:
I don't see the changes